docroot/sites/oniowa.uiowa.edu/modules/oniowa_core/oniowa_core.module

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/**
+ * @file
+ * Primary module hooks for Oniowa Core module.
+ */
+
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Implements hook_form_alter().
+ */
+function oniowa_core_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  // Use the form['#id'] over $form_id because it is more specific.
+  switch ($form['#id']) {
+    case 'views-exposed-form-events-page-upcoming':
+      $cid = 'sitenow_events:event:dates';
+      $data = \Drupal::cache()->get($cid);
+      if (!$data) {
+        $options = [];
+        $options[''] = t('Select a date');
+        $query = \Drupal::entityQuery('node');
+        $query->condition('type', 'event')
+          ->condition('status', 1)
+          ->condition('field_event_when.end_value', time(), '>')
+          ->sort('field_event_when', 'ASC');
+        $result = $query->execute();
+        if ($result) {
+          $nodes = \Drupal::entityTypeManager()
+            ->getStorage('node')
+            ->loadMultiple($result);
+          foreach ($nodes as $node) {
+            $date = $node->field_event_when->value;
+            if ($date) {
+              $day = date('l, F j', $date);
+              if (!isset($options[$date])) {
+                $options[$date] = $day;
+              }
+            }
+          }
+        }
+        // Cache the options, but set them to be invalidated
+        // should any events be CUDed.
+        $cache_tags = ['node_list:event'];
+        \Drupal::cache()->set(
+          $cid,
+          $options,
+          CacheBackendInterface::CACHE_PERMANENT,
+          $cache_tags
+        );
+      }
+      else {
+        $options = $data->data;
+      }
+      // Add an extra select list using the available date options.
+      $form['date_select'] = [
+        '#type' => 'select',
+        '#title' => t('Date'),
+        '#options' => $options,
+      ];
+      // Hide the date range fields, but leave them on the form.
+      // We'll still use them to avoid needing to alter
+      // the final views query.
+      $form['field_event_when_value']['#access'] = FALSE;
+      $form['field_event_when_end_value']['#access'] = FALSE;
+      array_unshift($form['#submit'], '_oniowa_core_date_select_list');
+      break;
+  }
+}
+
+/**
+ * Custom submit handler for upcoming events page.
+ */
+function _oniowa_core_date_select_list($form, &$form_state) {
```

Comment:
@joewhitsitt I seem to recall that we made some effort to move away from function names with an underscore prefix for reasons of incompatibility with future versions of PHPstan. Is this something we are trying to move away from?

Reply:
Not sure if we have made much effort, but that rule is on the horizon.

Reply:
Good info! I've updated to remove the underscore.